lib/libzfs/libzfs_sendrecv.c

```diff
@@ -1355,12 +1347,21 @@ dump_snapshot(zfs_handle_t *zhp, void *arg)
 
 	if (sdd->verbose) {
 		uint64_t size = 0;
-		(void) estimate_ioctl(zhp, sdd->prevsnap_obj,
-		    fromorigin, flags, &size);
+		char fromds[ZFS_MAX_DATASET_NAME_LEN];
 
-		send_print_verbose(fout, zhp->zfs_name,
-		    sdd->prevsnap[0] ? sdd->prevsnap : NULL,
-		    size, sdd->parsable);
+		if (sdd->prevsnap[0] != '\0') {
+			(void) strlcpy(fromds, zhp->zfs_name, sizeof (fromds));
+			*(strchr(fromds, '@') + 1) = '\0';
+			(void) strlcat(fromds, sdd->prevsnap, sizeof (fromds));
+		}
+		if (zfs_send_space(zhp, zhp->zfs_name,
```

Comment:
Although this doesn't look any worse than the original code, this seems to me like it might not print the correct thing when doing something like  `zfs send -i <3 snapshots ago> <latest snapshot>`. It seems like this will always just print the difference between the target snapshot and its previous snapshot.

Reply:
```
root@linux:~# zfs get written
NAME               PROPERTY  VALUE    SOURCE
testpool           written   24K      -
testpool/fs        written   0        -
testpool/fs@snap1  written   1.03M    -
testpool/fs@snap2  written   2.02M    -
testpool/fs@snap3  written   4.02M    -
testpool/fs@snap4  written   8.02M    -
root@linux:~# zfs send -v $POOLNAME/fs@snap1 | wc -c
full send of testpool/fs@snap1 estimated size is 1.01M
total estimated size is 1.01M
TIME        SENT   SNAPSHOT
1098368
root@linux:~# zfs send -v -i $POOLNAME/fs@snap1 $POOLNAME/fs@snap4 | wc -c
send from @snap1 to testpool/fs@snap4 estimated size is 14.0M
total estimated size is 14.0M
TIME        SENT   SNAPSHOT
14728216
root@linux:~# zfs send -v -i $POOLNAME/fs@snap2 $POOLNAME/fs@snap4 | wc -c
send from @snap2 to testpool/fs@snap4 estimated size is 12.0M
total estimated size is 12.0M
TIME        SENT   SNAPSHOT
12622616
root@linux:~# zfs send -v -i $POOLNAME/fs@snap3 $POOLNAME/fs@snap4 | wc -c
send from @snap3 to testpool/fs@snap4 estimated size is 8.02M
total estimated size is 8.02M
TIME        SENT   SNAPSHOT
8418016
root@linux:~# gdb -q zfs -ex 'b zfs_send_space' -ex "run send -vn -i $POOLNAME/fs@snap1 $POOLNAME/fs@snap4"
Reading symbols from zfs...done.
Function "zfs_send_space" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (zfs_send_space) pending.
Starting program: /usr/sbin/zfs send -vn -i testpool/fs@snap1 testpool/fs@snap4
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, zfs_send_space (zhp=0x62abe0, snapname=0x62abf0 "testpool/fs@snap4", from=0x7fffffffa660 "testpool/fs@snap1", flags=(unknown: 0), spacep=0x7fffffffa760) at libzfs_sendrecv.c:1019
warning: Source file is more recent than executable.
1019		libzfs_handle_t *hdl = zhp->zfs_hdl;
(gdb) fin
Run till exit from #0  zfs_send_space (zhp=0x62abe0, snapname=0x62abf0 "testpool/fs@snap4", from=0x7fffffffa660 "testpool/fs@snap1", flags=(unknown: 0), spacep=0x7fffffffa760) at libzfs_sendrecv.c:1019
0x00007ffff771efe0 in dump_snapshot (zhp=0x62abe0, arg=0x7fffffffe110) at libzfs_sendrecv.c:1357
1357			if (zfs_send_space(zhp, zhp->zfs_name,
Value returned is $1 = 0
(gdb) p size
$2 = 14717168
(gdb) 
```

Estimated size when sending from "testpool/fs@snap1" to "testpool/fs@snap4" is reported as 14M, which is 8M+4M+2M? Isn't this the expected result?

Reply:
Yeah I guess it is.... Not quite sure how that works.... I'll look into it for my own interest later today, but otherwise LGTM.